src/dag.rs

```diff
@@ -203,76 +144,122 @@ impl<G: GraphNode> Dag<G> {
+                .into_iter()
+                .collect::<Vec<&G>>()
+                .into_iter()
+                .rev(),
+        )
     }
 }
 
+// Possible states of a node during DFS.
+#[derive(Clone, PartialEq, Eq)]
+enum NodeState {
+    New,    // Initial state of every node.
+    Opened, // Node visited (pushed to the stack) but not yet closed.
+    Closed, // Node popped from the stack, all descendants visited.
+}
+
+// Struct that iterates over the nodes reachable by a set of given sources
+// in topological order (https://en.wikipedia.org/wiki/Topological_sorting),
+// that is, every node comes _after_ all its children (yes, techinically
+// this is reverse toposort, but here it kinda makes sense to call it that
+// since edges are meant to represent dependencies).
 #[derive(Clone)]
-pub struct BottomUp<'a, G: GraphNode> {
+pub struct TopologicalSort<'a, G: GraphNode> {
     dag: &'a Dag<G>,
-    visit_stack: Vec<usize>,
-    unvisited_roots: Vec<usize>,
+    dfs_stack: Vec<usize>,
+    node_state: Vec<NodeState>,
 }
 
-impl<'a, G: GraphNode> Iterator for BottomUp<'a, G> {
-    type Item = &'a G;
+impl<'a, G: GraphNode> TopologicalSort<'a, G> {
+    // The constructed instance will iterate over all nodes that are reachable
+    // from at least one of the given `sources`.
+    fn new(dag: &'a Dag<G>, sources: Vec<usize>) -> Self {
+        TopologicalSort {
+            dag,
+            dfs_stack: sources,
+            node_state: vec![NodeState::New; dag.nodes.len()],
+        }
+    }
 
-    fn next(&mut self) -> Option<&'a G> {
-        // I found the basic non-recursive DFS post-order algorithm here:
-        // https://codingots.medium.com/tree-traversal-without-recursion-221cbea6d004
-        // This is a translation of that, where "s1" is temp_stack and s2 is
-        // self.visit_stack. In that version there is only one root node but
-        // here we have several.
-        // First phase is to build up the stack of nodes to visit using
-        // temp_stack as an intermediate.
-        if self.visit_stack.is_empty() {
-            let mut temp_stack = vec![self.unvisited_roots.pop()?];
-            while let Some(cur_idx) = temp_stack.pop() {
-                self.visit_stack.push(cur_idx);
-                for child_idx in &self.dag.edges[cur_idx] {
-                    temp_stack.push(*child_idx);
+    // Private method that advances the iterator while checking for cycles.
+    //
+    // This is the iterative version of the DFS-based toposort implementation
+    // (https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search).
+    // Basically: do a normal DFS starting from the sources, and a node is
+    // appended to the toposort as soon as it is closed.
+    //
+    // If cycles are present, `next_or_error()` will find one eventually and
+    // return a `DagError::Cycle`. Subsequent calls will return `None`.
+    fn next_or_error(&mut self) -> Result<Option<&'a G>, DagError<G::NodeId>> {
```

Comment:
Normally the naming convention for "x, but it can fail" is `try_x` so this would be `try_next`

Reply:
Done.